.github/workflows/backend-cd.yml

```diff
@@ -0,0 +1,119 @@
+name: backend CD
+
+on:
+  push:
+    branches: [ main ]
+
+env:
+  REGISTRY: ghcr.io
+  IMAGE_NAME: ${{ github.repository }}/backend
+
+jobs:
+  build-and-deploy:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+
+    steps:
+      - name: checkout code
+        uses: actions/checkout@v4
+
+      - name: set up Go
+        uses: actions/setup-go@v5
+        with:
+          go-version: '1.23.5'
+          cache-dependency-path: backend/go.sum
+
+      - name: cache Go modules
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('backend/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+
+      - name: check Go formatting
+        working-directory: ./backend
+        run: |
+          unformatted=$(gofmt -l .)
+          if [ -n "$unformatted" ]; then
+            echo "The following files are not formatted:"
+            echo "$unformatted"
+            exit 1
+          fi
+
+      - name: download dependencies
+        working-directory: ./backend
+        run: go mod download
+
+      - name: run golangci-lint
+        uses: golangci/golangci-lint-action@v8
+        with:
+          version: latest
+          working-directory: ./backend
+          args: --timeout=5m
+
+      - name: run tests with race detector
+        working-directory: ./backend
+        run: go test -race -v ./...
+
+      - name: set up build variables
+        id: vars
+        run: |
+          COMMIT_SHA=$(git rev-parse --short HEAD)
```

Comment:
This would be available already through `GITHUB_SHA`.

Note that this can often be different from commit ID for `HEAD` you'd see in PR commit. In Azure DevOps pipelines, I've seen that it automatically merges `master` or `main` to PR before starting the pipeline execution. Not sure if that stupidity has propagated to GitHub Actions yet.